keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,309 @@
+
+BucketClass.Parameter key values are used to specify Secret name and namespace.  These may be considered of different degrees of granularity.
+
+- **Per Provisioner:** the Secret is used for all provisioning operations.  These Secrets should be injected directly in provisioner containers via [common Kubernetes patterns](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/).
+- **Per Operation/Bucket:** Secret’s data is passed to provisioners per operation or Bucket.  These Secrets should be defined as BucketClass parameters.  The keys should be clearly defined during automation design.
+
+This model should support both plain string values as well as “templates.”  Plain string values will allow admins to specify a particular Secret by name and namespace.  The key values should provide a means of defining *provisioner* credentials as well as *user* credentials.  User credentials may be required in cases where a user is being granted access to an existing bucket or when a user has permissions to create buckets themselves.
+
+As an example, the following key’s represent a minimun of what automation should support.  `provisioner-secret*` keys specify Secret to be used for bucket operations.  `user-secret-*` keys specify Secrets containing user identity information.
+
+```yaml
+cosi.io/provisioner-secret-name:
+cosi.io/provisioner-secret-namespace:
+cosi.io/user-secret-name:
+cosi.io/user-secret-namespace:
+```
+
+Bucket automation should support templating of Secret names and namespaces. For example:
+
+```yaml
+cosi.io/user-secret-name: "${bucket.name}"
+cosi.io/user-secret-namespace: "${bucket.namespace}-region-west"
+```
+
+> Note: Annotation key templates are not mentioned as of yet.  However, need for them may arise during automation design.
+
+## Custom Resource Definitions
+
+#### Bucket
+
+A user facing API object representing an object store bucket. Created by a user in their app's namespace. Once provisiong is complete, the `Bucket` is "bound" to the corresponding `BucketContent`. Binding is 1:1, meaning there is only one `BucketContent` per `Bucket` and vice-versa.
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
+  accessSecretName: [6]
+status:
+  bucketContentName: [7]
+  phase: [8]
+  conditions: 
+```
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.
+1. `protocol`: specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
```

Comment:
The `one of` is restrictive, and I'm not sure why there should be such restriction. This value should be opaque to anyone but a provisioner and a workload/client, so no need to restrict its value in the schema.

Reply:
"supportedProtocols" in bucketclass is the protocol list and check this. I am also confused for this, why not support one protocol for one bucketclass? and this restrict is not needed.

Reply:
From BucketClass description:
```
supportedProtocols: protocols the associated object store supports. Applied when matching Bucket to BucketClasses. Admins may specify all protocols a provisioner supports. Bucket.spec.protocol will be checked against the array prior to provisioning.
```
At one point in the design (and this may still be current @wlan0 ?) the bucketClass reference here would be validated to ensure it supported the requested protocol. If not then hard failure. That's my memory...

Reply:
bucketRequest is set to `InvalidArgument` phase, with a message describing that the bucketClass does not implement the requested protocol.